packages/inngest/src/components/execution/als.ts

```diff
@@ -0,0 +1,52 @@
+import { type Context } from "../../types.js";
+
+export interface AsyncContext {
+  ctx: Context.Any;
+}
+
+/**
+ * A type that represents a partial, runtime-agnostic interface of
+ * `AsyncLocalStorage`.
+ */
+type AsyncLocalStorageIsh = {
+  getStore: () => AsyncContext | undefined;
+  run: <R>(store: AsyncContext, fn: () => R) => R;
+};
+
+/**
+ * A local-only variable to store the async local storage instance.
+ */
+let als: Promise<AsyncLocalStorageIsh> | undefined;
+
+/**
+ * Retrieve the async context for the current execution.
+ */
+export const getAsyncCtx = async (): Promise<AsyncContext | undefined> => {
+  return getAsyncLocalStorage().then((als) => als.getStore());
+};
+
+/**
+ * Get a singleton instance of `AsyncLocalStorage` used to store and retrieve
+ * async context for the current execution.
+ */
+export const getAsyncLocalStorage = async (): Promise<AsyncLocalStorageIsh> => {
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises, no-async-promise-executor
+  als ??= new Promise<AsyncLocalStorageIsh>(async (resolve) => {
+    try {
+      const { AsyncLocalStorage } = await import("node:async_hooks");
```

Comment:
Bun, Deno, and Node all expose this as `node:async_hooks`.

Cloudflare Workers requires a compatibility flag of `node_compat` or `node_compat_v2`. A singular `node_als` should be possible but doesn't currently work with any version of Wrangler.

In the case that these flags have not been specified, this `AsyncLocalStorage` instance will essentially become a no-op; no store will ever be returned and `.run()` will just run the function given directly.

Reply:
Nice! 

On a side note, how many lines of code would it take you to do this in go!?

` als ??= new Promise<AsyncLocalStorageIsh>(async (resolve) => ...`